### ISSUE TITLE
help me please

### ISSUE THREAD

**Issue by TS-YinJie** [flexible-collision-library/fcl#339]
error: expected ‘}’ before ‘typedef’
error: expected ‘;’ before ‘typedef’
error: ‘S’ does not name a type
error: ‘S’ does not name a type
error: ‘Real’ does not name a type
error: ‘Real’ does not name a type
error: ‘Real’ does not name a type
error: ‘Real’ does not name a type
error: ‘Real’ does not name a type
error: ‘constants’ does not name a type
error: ‘constants’ does not name a type
error: expected declaration before ‘}’ token

such an error has made me crazy

**Comment by SeanCurtis-TRI** [comment]
To provide any help, we need a lot more. What OS, what compiler, can you show us the full error message? Etc.

**Comment by TS-YinJie** [comment]
thanks for your response. I am using ubuntu16.04 and ros-kinetic
when I use 'catkin_make' the error  appears
In file included from /usr/local/include/fcl/BV/RSS.h:41:0,
                 from /usr/local/include/fcl/ccd/motion_base.h:46,
                 from /usr/local/include/fcl/collision_object.h:45,
                 from /usr/local/include/fcl/octree.h:52,
                 from /home/yinjie/catkin_ws/src/path_planning/src/path_planning.cpp:40:
/usr/local/include/fcl/math/constants.h:131:1: error: expected primary-expression before ‘typedef’
 typedef typename detail::ScalarTrait<S>::type Real;
 ^
/usr/local/include/fcl/math/constants.h:131:1: error: expected ‘}’ before ‘typedef’
/usr/local/include/fcl/math/constants.h:131:1: error: expected ‘;’ before ‘typedef’
/usr/local/include/fcl/math/constants.h:134:18: error: ‘S’ does not name a type
 static constexpr S pi() { return S(3.141592653589793238462643383279502884197169
                  ^
/usr/local/include/fcl/math/constants.h:137:18: error: ‘S’ does not name a type
 static constexpr S phi() { return S(1.61803398874989484820458683436563811772030
                  ^
/usr/local/include/fcl/math/constants.h:145:8: error: ‘Real’ does not name a type
 static Real gjk_default_tolerance() {
        ^
/usr/local/include/fcl/math/constants.h:151:8: error: ‘Real’ does not name a type
 static Real eps() {
        ^
/usr/local/include/fcl/math/constants.h:160:8: error: ‘Real’ does not name a type
 static Real eps_78() {
        ^
/usr/local/include/fcl/math/constants.h:166:8: error: ‘Real’ does not name a type
 static Real eps_34() {
        ^
/usr/local/include/fcl/math/constants.h:172:8: error: ‘Real’ does not name a type
 static Real eps_12() {
        ^
/usr/local/include/fcl/math/constants.h:179:20: error: ‘constants’ does not name a type
 using constantsf = constants<float>;
                    ^
/usr/local/include/fcl/math/constants.h:180:20: error: ‘constants’ does not name a type
 using constantsd = constants<double>;
                    ^
/usr/local/include/fcl/math/constants.h:182:1: error: expected declaration before ‘}’ token
 } // namespace fcl

**Comment by TS-YinJie** [comment]
/usr/local/include/fcl/math/constants.h:131:1: error: expected primary-expression before ‘typedef’
 typedef typename detail::ScalarTrait<S>::type Real;
 ^
/usr/local/include/fcl/math/constants.h:131:1: error: expected ‘}’ before ‘typedef’
/usr/local/include/fcl/math/constants.h:131:1: error: expected ‘;’ before ‘typedef’
/usr/local/include/fcl/math/constants.h:134:18: error: ‘S’ does not name a type
 static constexpr S pi() { return S(3.141592653589793238462643383279502884197169
                  ^
/usr/local/include/fcl/math/constants.h:137:18: error: ‘S’ does not name a type
 static constexpr S phi() { return S(1.61803398874989484820458683436563811772030
                  ^
/usr/local/include/fcl/math/constants.h:145:8: error: ‘Real’ does not name a type
 static Real gjk_default_tolerance() {
        ^
/usr/local/include/fcl/math/constants.h:151:8: error: ‘Real’ does not name a type
 static Real eps() {
        ^
/usr/local/include/fcl/math/constants.h:160:8: error: ‘Real’ does not name a type
 static Real eps_78() {
        ^
/usr/local/include/fcl/math/constants.h:166:8: error: ‘Real’ does not name a type
 static Real eps_34() {
        ^
/usr/local/include/fcl/math/constants.h:172:8: error: ‘Real’ does not name a type
 static Real eps_12() {
        ^
/usr/local/include/fcl/math/constants.h:179:20: error: ‘constants’ does not name a type
 using constantsf = constants<float>;
                    ^
/usr/local/include/fcl/math/constants.h:180:20: error: ‘constants’ does not name a type
 using constantsd = constants<double>;
                    ^
/usr/local/include/fcl/math/constants.h:182:1: error: expected declaration before ‘}’ token
 } // namespace fcl

**Comment by TS-YinJie** [comment]
I copied the whole error, can you see it?

**Comment by wxmerkt** [comment]
@TS-YinJie have you set your workspace to be ``std=c++11``, e.g., in additional CMake args?

**Comment by TS-YinJie** [comment]
> This seems related to #330 which also doesn't list a solution. @TS-YinJie can you confirm this?

yes, it's the same as the #330

**Comment by TS-YinJie** [comment]
The solution to this  error  is still needed. 
Many people met this error, please help us!

**Comment by ana-GT** [comment]
@TS-YinJie  : I had the same error as you mentioned and I realized that the reason was that *my own project* was causing the issues: My project was calling FCL headers that change locations in the latest version, for instance:

1. Before: fcl/collision_object.h
2. Now: fcl/narrowphase/collision_object.h

In this example: Since I didn't update the location, the header was not coming from the right location....so  a mixture of old and new headers occurred, then some new headers were never called in account of having the same guads as some old headers being called erroneously. Point being: In my case, some important headers, such as "fcl/exports.h" were never called (this header defines FCL_EXPORT, if I remember right), hence I got all the stream of errors you also got.

In a nutshell: Before giving it another try, make sure that all the FCL files in the project/package you are trying to compile actually exist and are in the same location as in FCL 0.5 (and if they are not, update the location accordingly). That did it for me.

**Comment by SeanCurtis-TRI** [comment]
@ana-GT Thanks for fielding this (and #330). We really appreciate the contribution.
